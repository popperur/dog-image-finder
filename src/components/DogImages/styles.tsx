import styled from "styled-components";

export const DogImagesContainer = styled.div`
  padding: 40px;
`;

export const DogImage = styled.img`
  object-fit: cover;
  width: 100%;
  height: 40vh;
  min-height: 400px;
  border: white solid 10px;
  border-radius: 8px;
  aspect-ratio: 16/9;

  @media (max-width: 576px), (max-height: 500px) {
    height: auto;
    border-width: 7px;
  }
`;
