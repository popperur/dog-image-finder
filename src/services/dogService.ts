import axios from "axios";
import capitalize from "antd/lib/_util/capitalize";
import handleError from "utils/errorHandler.ts";

const BASE_URL = "https://dog.ceo/api";

type DogBreeds = Record<string, string[]>;

// Maps dog.ceo breed names
// from {spaniel:["japanese","irish","cocker"], komondor: []}
// to ["Cocker Spaniel", "Irish Spaniel", "Japanese Spaniel", "Komondor"]
function transformBreedNames(dogBreeds: DogBreeds) {
  const breedNames: string[] = [];
  for (const breedName in dogBreeds) {
    const subBreeds = dogBreeds[breedName];
    if (subBreeds.length > 0) {
      subBreeds.forEach((subBreed: string) =>
        breedNames.push(`${capitalize(subBreed)} ${capitalize(breedName)}`),
      );
    } else {
      breedNames.push(capitalize(breedName));
    }
  }
  breedNames.sort((a, b) => a.localeCompare(b));
  return breedNames;
}

async function getBreedNames() {
  return axios
    .get(`${BASE_URL}/breeds/list/all`)
    .then(response => {
      if (response.data.status !== "success") {
        console.error(
          "Breed names could not be fetched from the Dog CEO site.",
        );
        return [];
      }
      return transformBreedNames(response.data.message);
    })
    .catch(error => handleError(error, []));
}

async function getDogImageURLs(breedName: string, imageCount: number = 10) {
  if (!breedName || !breedName.trim()) {
    throw new Error("breedName is required and cannot be empty.");
  }

  let masterBreedName = breedName.toLowerCase();
  let subBreedName = "";
  const words = masterBreedName.split(" ");
  if (words.length > 1) {
    // first word is the sub, the rest is the master breed
    // examples: irish spaniel
    subBreedName = words.shift() || "";
    masterBreedName = words.join(" ");
  }

  const breedPath = masterBreedName + (subBreedName ? `/${subBreedName}` : "");
  return axios
    .get(`${BASE_URL}/breed/${breedPath}/images/random/${imageCount}`)
    .then(response => {
      if (response.data.status !== "success") {
        console.error(
          `Breed image URLs for ${breedName} could not be fetched from the Dog CEO site.`,
        );
        return [];
      }
      return response.data.message;
    })
    .catch(error => handleError(error, []));
}

export { getBreedNames, getDogImageURLs };
