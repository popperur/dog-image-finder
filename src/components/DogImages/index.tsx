import { Col, Row } from "antd";
import { DogImage, DogImagesContainer } from "components/DogImages/styles.tsx";
import { useEffect, useState } from "react";
import { getDogImageURLs } from "services/dogService.ts";

interface DogImagesProps {
  selectedBreedName: string;
}

const DISPLAY_DOG_IMAGE_COUNT = 9;

function DogImages({ selectedBreedName }: DogImagesProps) {
  const [dogImageURLs, setDogImageURLs] = useState([]);

  useEffect(() => {
    getDogImageURLs(selectedBreedName, DISPLAY_DOG_IMAGE_COUNT)
      .then(dogImageURLs => {
        setDogImageURLs(dogImageURLs);
      })
      .catch(error => {
        console.error(`Dog images could not be retrieved: ${error}`);
      });
  }, [selectedBreedName]);

  return (
    <DogImagesContainer>
      {dogImageURLs.length > 0 && (
        <Row gutter={[40, 40]}>
          {dogImageURLs.map((dogImageUrl, index) => (
            <Col key={index} xs={24} sm={24} md={8} lg={8} xl={8}>
              <DogImage
                src={dogImageUrl}
                alt={`Dog Image ${index + 1}`}
                loading="lazy"
              />
            </Col>
          ))}
        </Row>
      )}
    </DogImagesContainer>
  );
}

export default DogImages;
