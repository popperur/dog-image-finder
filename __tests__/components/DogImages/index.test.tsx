import { render, waitFor, screen } from "@testing-library/react";
import { describe, it, expect, beforeEach, vi, afterEach } from "vitest";
import DogImages from "components/DogImages";
import { getDogImageURLs } from "services/dogService.ts";

beforeEach(() => {
  vi.mock("services/dogService", () => ({
    getDogImageURLs: vi.fn().mockResolvedValue(["image1.jpg", "image2.jpg"]),
  }));
});
afterEach(() => {
  vi.clearAllMocks();
});

describe("DogImages component", () => {
  describe("snapshot", () => {
    it("matches the snapshot", async () => {
      const { asFragment } = render(<DogImages selectedBreedName="komondor" />);

      await waitFor(() => {
        expect(asFragment()).toMatchSnapshot();
      });
    });
  });

  describe("dog images", () => {
    it("fetches the dog image URLs", async () => {
      render(<DogImages selectedBreedName="Komondor" />);

      await waitFor(() => {
        expect(getDogImageURLs).toHaveBeenCalledWith("Komondor", 24);
      });
    });

    it("renders correctly with dog images for the selected breed", async () => {
      render(<DogImages selectedBreedName="Komondor" />);

      await waitFor(() => {
        const images = screen.getAllByRole("img");
        expect(images).toHaveLength(2);
        expect(images[0]).toHaveAttribute("src", "image1.jpg");
        expect(images[0]).toHaveAttribute("alt", "Dog Image");
        expect(images[1]).toHaveAttribute("src", "image2.jpg");
        expect(images[1]).toHaveAttribute("alt", "Dog Image");
      });
    });

    it("handles API error gracefully and logs to console", async () => {
      const consoleErrorMock = vi
        .spyOn(console, "error")
        .mockImplementation(() => {});
      const mockedGetDogImageURLs = vi.mocked(getDogImageURLs);
      mockedGetDogImageURLs.mockRejectedValue(new Error("Network Error"));

      render(<DogImages selectedBreedName="Komondor" />);

      await waitFor(() => {
        expect(consoleErrorMock).toHaveBeenCalledWith(
          "Dog images could not be retrieved: Error: Network Error",
        );
        expect(screen.queryAllByRole("img")).toHaveLength(0);
      });

      consoleErrorMock.mockRestore();
    });
  });
});
