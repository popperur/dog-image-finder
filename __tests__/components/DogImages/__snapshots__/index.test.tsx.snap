// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DogImages component > Snapshot > matches the snapshot 1`] = `
<DocumentFragment>
  .c0 {
  padding: 40px;
}

<div
    class="c0"
  />
</DocumentFragment>
`;
